homeassistant/components/template/binary_sensor.py

```diff
@@ -54,96 +64,151 @@
 
-        sensors.append(
-            BinarySensorTemplate(
-                hass,
-                device,
-                friendly_name,
-                device_class,
-                value_template,
-                icon_template,
-                entity_picture_template,
-                availability_template,
-                entity_ids,
-                delay_on,
-                delay_off,
-                attribute_templates,
-            )
-        )
+class BinarySensorStorageCollection(collection.StorageCollection):
+    """Input storage based collection."""
 
-    async_add_entities(sensors)
+    CREATE_SCHEMA = vol.Schema(CREATE_FIELDS)
+    UPDATE_SCHEMA = vol.Schema(UPDATE_FIELDS)
 
+    async def _process_create_data(self, data: typing.Dict) -> typing.Dict:
+        """Validate the config is valid."""
+        return self.CREATE_SCHEMA(data)
 
-class BinarySensorTemplate(BinarySensorEntity):
-    """A virtual binary sensor that triggers from another sensor."""
+    @callback
+    def _get_suggested_id(self, info: typing.Dict) -> str:
+        """Suggest an ID based on the config."""
+        return info[CONF_FRIENDLY_NAME]
+
+    async def _update_data(self, data: dict, update_data: typing.Dict) -> typing.Dict:
+        """Return a new updated data object."""
+        self.UPDATE_SCHEMA(update_data)
+        return {**data, **update_data}
+
+
+def init_config(hass, device: str, config: typing.Dict):
+    """Initialise templates and entities to be watched."""
+    templates = {
+        CONF_VALUE_TEMPLATE: config.get(CONF_VALUE_TEMPLATE),
+        CONF_ICON_TEMPLATE: config.get(CONF_ICON_TEMPLATE),
+        CONF_ENTITY_PICTURE_TEMPLATE: config.get(CONF_ENTITY_PICTURE_TEMPLATE),
+        CONF_AVAILABILITY_TEMPLATE: config.get(CONF_AVAILABILITY_TEMPLATE),
+    }
+
+    attribute_templates = config.get(CONF_ATTRIBUTE_TEMPLATES, {})
 
-    def __init__(
-        self,
-        hass,
+    initialise_templates(hass, templates, attribute_templates)
+    config[CONF_ENTITY_ID] = extract_entities(
         device,
-        friendly_name,
-        device_class,
-        value_template,
-        icon_template,
-        entity_picture_template,
-        availability_template,
-        entity_ids,
-        delay_on,
-        delay_off,
+        "binary sensor",
+        config.get(CONF_ENTITY_ID),
+        templates,
         attribute_templates,
-    ):
+    )
+
+    return config
+
+
+class BinarySensorTemplate(BinarySensorEntity):
+    """A virtual binary sensor that triggers from another sensor."""
+
+    def __init__(self, hass: HomeAssistantType, config: ConfigType):
         """Initialize the Template binary sensor."""
         self.hass = hass
-        self.entity_id = async_generate_entity_id(ENTITY_ID_FORMAT, device, hass=hass)
-        self._name = friendly_name
-        self._device_class = device_class
-        self._template = value_template
+        self.editable = False
+        self.entity_id = async_generate_entity_id(
```

Comment:
We should not do this, this gives horrible entity IDs. Since there is a unique ID, it will work just fine to let it be based on name.

Reply:
This is the existing entity_id code. Are you meaning something like `f"{DOMAIN}.{config[CONF_ID]"` instead?

Reply:
Aah. In that case we should only do that if the entity is based on YAML. 

If we don't define a name, Home Assistant will automatically generate one based on the name and persist it in the entity registry (because of the unique ID)